src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,211 @@ internal class DiagnosticComputer
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetHighPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Step 1:
+            //  - Create the core 'computeTask' for computing diagnostics.
+            var computeTask = Task.Run(
+                () => GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken),
+                cancellationToken);
+
+            // Step 2:
+            //  - Add this computeTask to the list of currently executing high priority tasks.
+            //    This list of high priority tasks is used in 'GetNormalPriorityDiagnosticsAsync'
+            //    method to ensure that any new or cancelled normal priority task waits for all
+            //    the executing high priority tasks before starting its execution.
+            //  - Note that it is critical to do this step prior to Step 3 below to ensure that
+            //    any canceled normal priority tasks in Step 3 do not resume execution prior to
+            //    completion of this high priority computeTask. 
+            lock (s_gate)
+            {
+                Debug.Assert(!s_highPriorityComputeTasks.Contains(computeTask));
+                s_highPriorityComputeTasks.Add(computeTask);
+            }
+
+            try
+            {
+                // Step 3:
+                //  - Force cancellation of all the executing normal priority tasks
+                //    to minimize resource and CPU contention between normal priority tasks
+                //    and the high priority computeTask in Step 4 below.
+                CancelNormalPriorityTasks();
+
+                // Step 4:
+                //  - Execute the core 'computeTask' for diagnostic computation.
+                return await computeTask.ConfigureAwait(false);
+            }
+            finally
+            {
+                // Step 5:
+                //  - Remove the 'computeTask' from the list of current executing high priority tasks.
+                lock (s_gate)
+                {
+                    var removed = s_highPriorityComputeTasks.Remove(computeTask);
+                    Debug.Assert(removed);
+                }
+            }
+
+            static void CancelNormalPriorityTasks()
+            {
+                ImmutableArray<CancellationTokenSource> cancellationTokenSourcesToCancel;
+                lock (s_gate)
+                {
+                    cancellationTokenSourcesToCancel = s_cancellationSourcesForNormalPriorityComputeTasks.ToImmutableArrayOrEmpty();
+                }
+
+                foreach (var cancellationTokenSource in cancellationTokenSourcesToCancel)
```

Comment:
note to self: this coudl run in the lock, and then avoid the object-disposed-exception.  but there is merit in having the locking portions be a minimal as possible.

Reply:
> note to self: this coudl run in the lock, and then avoid the object-disposed-exception. but there is merit in having the locking portions be a minimal as possible.

I also considered this, but chose the latter